File: lite/examples/pose_estimation/android/app/src/main/java/org/tensorflow/lite/examples/poseestimation/ml/MoveNetMultiPose.kt
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================
*/

package org.tensorflow.lite.examples.poseestimation.ml

import android.content.Context
import android.graphics.Bitmap
import android.graphics.PointF
import android.graphics.RectF
import android.os.SystemClock
import org.tensorflow.lite.DataType
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.examples.poseestimation.data.BodyPart
import org.tensorflow.lite.examples.poseestimation.data.Device
import org.tensorflow.lite.examples.poseestimation.data.KeyPoint
import org.tensorflow.lite.examples.poseestimation.data.Person
import org.tensorflow.lite.examples.poseestimation.tracker.*
import org.tensorflow.lite.gpu.GpuDelegate
import org.tensorflow.lite.support.common.FileUtil
import org.tensorflow.lite.support.image.ImageOperator
import org.tensorflow.lite.support.image.ImageProcessor
import org.tensorflow.lite.support.image.TensorImage
import org.tensorflow.lite.support.image.ops.ResizeOp
import org.tensorflow.lite.support.image.ops.ResizeWithCropOrPadOp
import org.tensorflow.lite.support.tensorbuffer.TensorBuffer
import kotlin.math.ceil

class MoveNetMultiPose(
    private val interpreter: Interpreter,
    private val type: Type,
    private val gpuDelegate: GpuDelegate?,
) : PoseDetector {
    private val outputShape = interpreter.getOutputTensor(0).shape()
    private val inputShape = interpreter.getInputTensor(0).shape()
    private var imageWidth: Int = 0
    private var imageHeight: Int = 0
    private var targetWidth: Int = 0
    private var targetHeight: Int = 0
    private var scaleHeight: Int = 0
    private var scaleWidth: Int = 0
    private var lastInferenceTimeNanos: Long = -1
    private var tracker: AbstractTracker? = null

    companion object {
        private const val DYNAMIC_MODEL_TARGET_INPUT_SIZE = 256
        private const val SHAPE_MULTIPLE = 32.0
        private const val DETECTION_THRESHOLD = 0.11
        private const val DETECTION_SCORE_INDEX = 55
        private const val BOUNDING_BOX_Y_MIN_INDEX = 51
        private const val BOUNDING_BOX_X_MIN_INDEX = 52
        private const val BOUNDING_BOX_Y_MAX_INDEX = 53
        private const val BOUNDING_BOX_X_MAX_INDEX = 54
        private const val KEYPOINT_COUNT = 17
        private const val OUTPUTS_COUNT_PER_KEYPOINT = 3
        private const val CPU_NUM_THREADS = 4

        // allow specifying model type.
        fun create(
            context: Context,
            device: Device,
            type: Type,
        ): MoveNetMultiPose {
            val options = Interpreter.Options()
            var gpuDelegate: GpuDelegate? = null
            when (device) {
                Device.CPU -> {
                    options.setNumThreads(CPU_NUM_THREADS)
                }
                Device.GPU -> {
                    // only fixed model support Gpu delegate option.
                    if (type == Type.Fixed) {
                        gpuDelegate = GpuDelegate()
                        options.addDelegate(gpuDelegate)
                    }
                }
                else -> {
                    // nothing to do
                }
            }
            return MoveNetMultiPose(
                Interpreter(
                    FileUtil.loadMappedFile(
                        context,
                        if (type == Type.Dynamic)
                            "movenet_multipose_fp16.tflite" else ""
                        //@TODO: (khanhlvg) Add support for fixed shape model if it's released.
                    ), options
                ), type, gpuDelegate
            )
        }
    }

    /**
     * Convert x and y coordinates ([0-1]) returns from the TFlite model
     * to the coordinates corresponding to the input image.
     */
    private fun resizeKeypoint(x: Float, y: Float): PointF {
        return PointF(resizeX(x), resizeY(y))
    }

    private fun resizeX(x: Float): Float {
        return if (imageWidth > imageHeight) {
            val ratioWidth = imageWidth.toFloat() / targetWidth
            x * targetWidth * ratioWidth
        } else {
            val detectedWidth =
                if (type == Type.Dynamic) targetWidth else inputShape[2]
            val paddingWidth = detectedWidth - scaleWidth
            val ratioWidth = imageWidth.toFloat() / scaleWidth
            (x * detectedWidth - paddingWidth / 2f) * ratioWidth
        }
    }

    private fun resizeY(y: Float): Float {
        return if (imageWidth > imageHeight) {
            val detectedHeight =
                if (type == Type.Dynamic) targetHeight else inputShape[1]
            val paddingHeight = detectedHeight - scaleHeight
            val ratioHeight = imageHeight.toFloat() / scaleHeight
            (y * detectedHeight - paddingHeight / 2f) * ratioHeight
        } else {
            val ratioHeight = imageHeight.toFloat() / targetHeight
            y * targetHeight * ratioHeight
        }
    }

    /**
     * Prepare input image for detection
     */
    private fun processInputTensor(bitmap: Bitmap): TensorImage {
        imageWidth = bitmap.width
        imageHeight = bitmap.height

        // if model type is fixed. get input size from input shape.
        val inputSizeHeight =
            if (type == Type.Dynamic) DYNAMIC_MODEL_TARGET_INPUT_SIZE else inputShape[1]
        val inputSizeWidth =
            if (type == Type.Dynamic) DYNAMIC_MODEL_TARGET_INPUT_SIZE else inputShape[2]

        val resizeOp: ImageOperator
        if (imageWidth > imageHeight) {
            val scale = inputSizeWidth / imageWidth.toFloat()
            targetWidth = inputSizeWidth
            scaleHeight = ceil(imageHeight * scale).toInt()
            targetHeight = (ceil((scaleHeight / SHAPE_MULTIPLE)) * SHAPE_MULTIPLE).toInt()
            resizeOp = ResizeOp(scaleHeight, targetWidth, ResizeOp.ResizeMethod.BILINEAR)
        } else {
            val scale = inputSizeHeight / imageHeight.toFloat()
            targetHeight = inputSizeHeight
            scaleWidth = ceil(imageWidth * scale).toInt()
            targetWidth = (ceil((scaleWidth / SHAPE_MULTIPLE)) * SHAPE_MULTIPLE).toInt()
            resizeOp = ResizeOp(targetHeight, scaleWidth, ResizeOp.ResizeMethod.BILINEAR)
        }

        val resizeWithCropOrPad = if (type == Type.Dynamic) ResizeWithCropOrPadOp(
            targetHeight,
            targetWidth
        ) else ResizeWithCropOrPadOp(
            inputSizeHeight,
            inputSizeWidth
        )
        val imageProcessor = ImageProcessor.Builder().apply {
            add(resizeOp)
            add(resizeWithCropOrPad)
        }.build()
        val tensorImage = TensorImage(DataType.UINT8)
        tensorImage.load(bitmap)
        return imageProcessor.process(tensorImage)
    }

    /**
     * Run tracker (if available) and process the output.
     */
    private fun postProcess(modelOutput: FloatArray): List<Person> {
        val persons = mutableListOf<Person>()
        for (idx in modelOutput.indices step outputShape[2]) {
            val personScore = modelOutput[idx + DETECTION_SCORE_INDEX]
            if (personScore < DETECTION_THRESHOLD) continue
            val positions = modelOutput.copyOfRange(idx, idx + 51)
            val keyPoints = mutableListOf<KeyPoint>()
            for (i in 0 until KEYPOINT_COUNT) {
                val y = positions[i * OUTPUTS_COUNT_PER_KEYPOINT]
                val x = positions[i * OUTPUTS_COUNT_PER_KEYPOINT + 1]
                val score = positions[i * OUTPUTS_COUNT_PER_KEYPOINT + 2]
                keyPoints.add(KeyPoint(BodyPart.fromInt(i), PointF(x, y), score))
            }
            val yMin = modelOutput[idx + BOUNDING_BOX_Y_MIN_INDEX]
            val xMin = modelOutput[idx + BOUNDING_BOX_X_MIN_INDEX]
            val yMax = modelOutput[idx + BOUNDING_BOX_Y_MAX_INDEX]
            val xMax = modelOutput[idx + BOUNDING_BOX_X_MAX_INDEX]
            val boundingBox = RectF(xMin, yMin, xMax, yMax)
            persons.add(
                Person(
                    keyPoints = keyPoints,
                    boundingBox = boundingBox,
                    score = personScore
                )
            )
        }

        if (persons.isEmpty()) return emptyList()

        if (tracker == null) {
            persons.forEach {
                it.keyPoints.forEach { key ->
                    key.coordinate = resizeKeypoint(key.coordinate.x, key.coordinate.y)
                }
            }
            return persons
        } else {
            val trackPersons = mutableListOf<Person>()
            tracker?.apply(persons, System.currentTimeMillis() * 1000)?.forEach {
                val resizeKeyPoint = mutableListOf<KeyPoint>()
                it.keyPoints.forEach { key ->
                    resizeKeyPoint.add(
                        KeyPoint(
                            key.bodyPart,
                            resizeKeypoint(key.coordinate.x, key.coordinate.y),
                            key.score
                        )
                    )
                }

                var resizeBoundingBox: RectF? = null
                it.boundingBox?.let { boundingBox ->
                    resizeBoundingBox = RectF(
                        resizeX(boundingBox.left),
                        resizeY(boundingBox.top),
                        resizeX(boundingBox.right),
                        resizeY(boundingBox.bottom)
                    )
                }
                trackPersons.add(Person(it.id, resizeKeyPoint, resizeBoundingBox, it.score))
            }
            return trackPersons
        }
    }

    /**
     * Create and set tracker.
     */
    fun setTracker(trackerType: TrackerType) {
        tracker = when (trackerType) {
            TrackerType.BOUNDING_BOX -> {
                BoundingBoxTracker()
            }
            TrackerType.KEYPOINTS -> {
                KeyPointsTracker()
            }
            TrackerType.OFF -> {
                null
            }
        }
    }

    /**
     * Run TFlite model and Returns a list of "Person" corresponding to the input image.
     */
    override fun estimatePoses(bitmap: Bitmap): List<Person> {
        val inferenceStartTimeNanos = SystemClock.elapsedRealtimeNanos()
        val inputTensor = processInputTensor(bitmap)
        val outputTensor = TensorBuffer.createFixedSize(outputShape, DataType.FLOAT32)

        // if model is dynamic, resize input before run interpreter
        if (type == Type.Dynamic) {
            val inputShape = intArrayOf(1).plus(inputTensor.tensorBuffer.shape)
            interpreter.resizeInput(0, inputShape, true)
            interpreter.allocateTensors()
        }
        interpreter.run(inputTensor.buffer, outputTensor.buffer.rewind())

        val processedPerson = postProcess(outputTensor.floatArray)
        lastInferenceTimeNanos =
            SystemClock.elapsedRealtimeNanos() - inferenceStartTimeNanos
        return processedPerson
    }

    override fun lastInferenceTimeNanos(): Long = lastInferenceTimeNanos

    /**
     * Close all resources when not in use.
     */
    override fun close() {
        gpuDelegate?.close()
        interpreter.close()
        tracker = null
    }
}

enum class Type {
    Dynamic, Fixed
}

enum class TrackerType {
    OFF, BOUNDING_BOX, KEYPOINTS
}

File: lite/examples/pose_estimation/ios/PoseEstimation/ML/Models/MoveNetMultiPose.swift
// Copyright 2021 The TensorFlow Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================
//

import Accelerate
import CoreImage
import TensorFlowLite

class MoveNetMultiPose: PoseEstimator {
    private let interpreter: Interpreter
    // TensorFlow Lite `Tensor` of model input and output.
    private var inputTensor: Tensor
    private var outputTensor: Tensor?

    private let modelType: MultiPoseType
    private var imageWidth = 0
    private var imageHeight = 0
    private var targetWidth: Int = 0
    private var targetHeight: Int = 0
    private var scaleHeight: Int = 0
    private var scaleWidth: Int = 0
    
    private var lastInferenceTime: TimeInterval = 0

    // Model config
    private var torsoExpansionRatio = 1.9
    private var bodyExpandsionRatio = 1.2
    private let imageMean: Float = 0
    private let imageStd: Float = 1
    private let minCropKeyPointScore: Float32 = 0.2
    private var cropRegion: RectF?
    private var isProcessing = false
    
    private let dynamicModelTargetInputSize = 256
    private let shapeMultiple = 32.0
    private let detectionThreshold: Float32 = Constants.minimumScore
    private let detectionScoreIndex = 55
    private let boundingBoxYMinIndex = 51
    private let boundingBoxXMinIndex = 52
    private let boundingBoxYMaxIndex = 53
    private let boundingBoxXMaxIndex = 54
    private let keypointCount = 17
    private let outputsCountPerKeypoint = 3
    private let cpuNumThreads = 4

    private let movenetMultiposeLightning = FileInfo(name: "movenet_multipose_fp16", ext: "tflite")

    init(threadCount: Int, delegate: Delegates, modelType: MultiPoseType) throws {
        self.modelType = modelType
        var modelPath: String?
        switch modelType {
        case .dynamic:
            modelPath = Bundle.main.path(
                forResource: movenetMultiposeLightning.name,
                ofType: movenetMultiposeLightning.ext)
        case .fixed:
            modelPath = ""
        }
        guard let modelPath = modelPath else {
            fatalError("Failed to load the model file.")
        }

        // Specify the options for the `Interpreter`.
        var options = Interpreter.Options()
        // Specify the delegates for the `Interpreter`.
        var delegates: [Delegate]?
        switch delegate {
        case .gpu:
            options.threadCount = threadCount
            if modelType == .fixed {
                delegates = [MetalDelegate()]
            }
        case .npu:
            delegates = nil
        case .cpu:
            options.threadCount = threadCount
            delegates = nil
        }
        // Create the `Interpreter`.
        interpreter = try Interpreter(modelPath: modelPath, options: options, delegates: delegates)

        // Initialize input and output `Tensor`s.
        // Allocate memory for the model's input `Tensor`s.
        try interpreter.allocateTensors()

        // Get allocated input and output `Tensor`s.
        inputTensor = try interpreter.input(at: 0)
        outputTensor = try? interpreter.output(at: 0)
    }

    func estimatePoses(on pixelBuffer: CVPixelBuffer) throws -> ([Person], Times) {
        // Check if this MoveNet instance is already processing a video frame.
        // Return an empty detection result if it's currently busy.

        guard !isProcessing else {
            throw PoseEstimationError.modelBusy
        }
        isProcessing = true
        defer {
            isProcessing = false
        }

        // Start times of each process.
        let preprocessingStartTime: Date
        let inferenceStartTime: Date
        let postprocessingStartTime: Date

        // Processing times in seconds.
        let preprocessingTime: TimeInterval
        let inferenceTime: TimeInterval
        let postprocessingTime: TimeInterval

        preprocessingStartTime = Date()
        guard let rgbModel = preprocess(pixelBuffer) else {
            os_log("Preprocessing failed.", type: .error)
            throw PoseEstimationError.preprocessingFailed
        }
        preprocessingTime = Date().timeIntervalSince(preprocessingStartTime)

        // Run inference with the TFLite model
        inferenceStartTime = Date()
        do {
            if modelType == .dynamic {
                var shape = rgbModel.shape
                shape.insert(1, at: 0)
                let tensorShape: Tensor.Shape = Tensor.Shape(shape)
                try interpreter.resizeInput(at: 0, to: tensorShape)
                try interpreter.allocateTensors()
            }
            // Copy the initialized `Data` to the input `Tensor`.
            try interpreter.copy(rgbModel.imageData, toInputAt: 0)

            // Run inference by invoking the `Interpreter`.
            try interpreter.invoke()
            // Get the output `Tensor` to process the inference results.
            outputTensor = try interpreter.output(at: 0)
        } catch {
            os_log(
                "Failed to invoke the interpreter with error: %s", type: .error,
                error.localizedDescription)
            throw PoseEstimationError.inferenceFailed
        }
        inferenceTime = Date().timeIntervalSince(inferenceStartTime)

        postprocessingStartTime = Date()
        guard let outputTensor = outputTensor, let result = postprocess(imageSize: pixelBuffer.size, modelOutput: outputTensor) else {
            os_log("Postprocessing failed.", type: .error)
            throw PoseEstimationError.postProcessingFailed
        }
        postprocessingTime = Date().timeIntervalSince(postprocessingStartTime)

        let times = Times(
            preprocessing: preprocessingTime,
            inference: inferenceTime,
            postprocessing: postprocessingTime)
        return (result, times)
    }

    // MARK: - Private functions to run the model

    /// Preprocesses given rectangle image to be `Data` of desired size by cropping and resizing it.
    ///
    /// - Parameters:
    ///   - of: Input image to crop and resize.
    /// - Returns: The cropped and resized image. `nil` if it can not be processed.
    private func preprocess(_ pixelBuffer: CVPixelBuffer) -> InputRGBModel? {
        let sourcePixelFormat = CVPixelBufferGetPixelFormatType(pixelBuffer)
        assert(
            sourcePixelFormat == kCVPixelFormatType_32BGRA
                || sourcePixelFormat == kCVPixelFormatType_32ARGB)
        // Resize `targetSquare` of input image to `modelSize`.
        let dimensions = inputTensor.shape.dimensions
        let inputWidth = modelType == .dynamic ? dynamicModelTargetInputSize : dimensions[1]
        let inputHeight = modelType == .dynamic ? dynamicModelTargetInputSize : dimensions[2]
        let imageWidth = pixelBuffer.size.width
        let imageHeight = pixelBuffer.size.height

        let cropRegion = self.cropRegion ??
            initialCropRegion(imageWidth: imageWidth, imageHeight: imageHeight)
        self.cropRegion = cropRegion

        let rectF = RectF(
            left: cropRegion.left * imageWidth,
            top: cropRegion.top * imageHeight,
            right: cropRegion.right * imageWidth,
            bottom: cropRegion.bottom * imageHeight)

        // Detect region
        let modelSize = CGSize(width: inputWidth, height: inputHeight)
        guard let thumbnail = pixelBuffer.cropAndResize(fromRect: rectF.rect, toSize: modelSize) else {
            return nil
        }

        // Remove the alpha component from the image buffer to get the initialized `Data`.
        guard
            let inputData = thumbnail.rgbData(
                isModelQuantized: inputTensor.dataType == .uInt8, imageMean: imageMean, imageStd: imageStd)
        else {
            os_log("Failed to convert the image buffer to RGB data.", type: .error)
            return nil
        }
        let width = inputWidth
        let height = inputHeight
        // rgb so channelCount = 3
        let destinationChannelCount = 3
        return InputRGBModel(imageData: inputData, shape: [width, height, destinationChannelCount])
    }

    /// Postprocesses the output `Tensor` of TFLite model to the `Person` type
    ///
    /// - Parameters:
    ///   - imageSize: Size of the input image.
    ///   - modelOutput: Output tensor from the TFLite model.
    /// - Returns: Postprocessed `Person`. `nil` if it can not be processed.
    private func postprocess(imageSize: CGSize, modelOutput: Tensor) -> [Person]? {
        let imageWidth = imageSize.width
        let imageHeight = imageSize.height

        let output = modelOutput.data.toArray(type: Float32.self)
        let dimensions = modelOutput.shape.dimensions
        let numKeyPoints = dimensions[2]
        let inputWidth = (modelType == .dynamic) ? CGFloat(dynamicModelTargetInputSize) : CGFloat(inputTensor.shape.dimensions[1])
        let inputHeight = (modelType == .dynamic) ? CGFloat(dynamicModelTargetInputSize) : CGFloat(inputTensor.shape.dimensions[2])
        let cropRegion = self.cropRegion ??
            initialCropRegion(imageWidth: imageWidth, imageHeight: imageHeight)

        let minX: CGFloat = cropRegion.left * imageWidth
        let minY: CGFloat = cropRegion.top * imageHeight
        let widthRatio = (cropRegion.width * imageWidth / inputWidth)
        let heightRatio = (cropRegion.height * imageHeight / inputHeight)
        var persons: [Person] = []
        // Translate the coordinates from the model output's [0..1] back to that of
        // the input image output.count
        for index in stride(from: 0, to: output.count, by: numKeyPoints) {
            let personScore = output[index + detectionScoreIndex]
            if personScore < detectionThreshold {
                continue
            }
            let positions: [Float32] = Array(output[index ..< index + 51])
            var totalScoreSum: Float32 = 0
            var keyPoints: [KeyPoint] = []
            for idx in 0 ..< keypointCount {
                let x = ((CGFloat(positions[idx * outputsCountPerKeypoint + 1]) * inputWidth) * widthRatio) + minX
                let y = ((CGFloat(positions[idx * outputsCountPerKeypoint + 0]) * inputHeight) * heightRatio) + minY
                let score = positions[idx * outputsCountPerKeypoint + 2]
                totalScoreSum += score
                let keyPoint = KeyPoint(
                    bodyPart: BodyPart.allCases[idx], coordinate: CGPoint(x: x, y: y), score: score)
                keyPoints.append(keyPoint)
            }
            // Calculates total confidence score of each key position.
            let totalScore = totalScoreSum / Float32(numKeyPoints)

            // Make `Person` from `keypoints'. Each point is adjusted to the coordinate of the input image.
            let person = Person(keyPoints: keyPoints, score: totalScore)
            persons.append(person)
        }
        return persons
    }

    // MARK: MoveNet's smart cropping logic

    /// Determines the region to crop the image for the model to run inference on.
    /// The algorithm uses the detected joints from the previous frame to estimate
    /// the square region that encloses the full body of the target person and
    /// centers at the midpoint of two hip joints. The crop size is determined by
    /// the distances between each joints and the center point.
    /// When the model is not confident with the four torso joint predictions, the
    /// function returns a default crop which is the full image padded to square.
    private func nextFrameCropRegion(keyPoints: [KeyPoint], imageWidth: CGFloat, imageHeight: CGFloat)
        -> RectF
    {
        let targetKeyPoints = keyPoints.map { keyPoint in
            KeyPoint(bodyPart: keyPoint.bodyPart,
                     coordinate: CGPoint(x: keyPoint.coordinate.x, y: keyPoint.coordinate.y),
                     score: keyPoint.score)
        }
        if torsoVisible(keyPoints) {
            let centerX =
                (targetKeyPoints[BodyPart.leftHip.position].coordinate.x
                    + targetKeyPoints[BodyPart.rightHip.position].coordinate.x) / 2.0
            let centerY =
                (targetKeyPoints[BodyPart.leftHip.position].coordinate.y
                    + targetKeyPoints[BodyPart.rightHip.position].coordinate.y) / 2.0

            let torsoAndBodyDistances =
                determineTorsoAndBodyDistances(
                    keyPoints: keyPoints, targetKeyPoints: targetKeyPoints, centerX: centerX, centerY: centerY)

            let list = [
                torsoAndBodyDistances.maxTorsoXDistance * CGFloat(torsoExpansionRatio),
                torsoAndBodyDistances.maxTorsoYDistance * CGFloat(torsoExpansionRatio),
                torsoAndBodyDistances.maxBodyXDistance * CGFloat(bodyExpandsionRatio),
                torsoAndBodyDistances.maxBodyYDistance * CGFloat(bodyExpandsionRatio),
            ]

            var cropLengthHalf = list.max() ?? 0.0
            let tmp: [CGFloat] = [
                centerX, CGFloat(imageWidth) - centerX, centerY, CGFloat(imageHeight) - centerY,
            ]
            cropLengthHalf = min(cropLengthHalf, tmp.max() ?? 0.0)
            let cropCornerY = centerY - cropLengthHalf
            let cropCornerX = centerX - cropLengthHalf
            if cropLengthHalf > (CGFloat(max(imageWidth, imageHeight)) / 2.0) {
                return initialCropRegion(imageWidth: imageWidth, imageHeight: imageHeight)
            } else {
                let cropLength = cropLengthHalf * 2
                return RectF(
                    left: max(cropCornerX, 0) / imageWidth,
                    top: max(cropCornerY, 0) / imageHeight,
                    right: min((cropCornerX + cropLength) / imageWidth, 1),
                    bottom: min((cropCornerY + cropLength) / imageHeight, 1))
            }
        } else {
            return initialCropRegion(imageWidth: imageWidth, imageHeight: imageHeight)
        }
    }

    /// Defines the default crop region.
    /// The function provides the initial crop region (pads the full image from both
    /// sides to make it a square image) when the algorithm cannot reliably determine
    /// the crop region from the previous frame.
    private func initialCropRegion(imageWidth: CGFloat, imageHeight: CGFloat) -> RectF {
        var xMin: CGFloat
        var yMin: CGFloat
        var width: CGFloat
        var height: CGFloat
        if imageWidth > imageHeight {
            height = 1
            width = imageHeight / imageWidth
            yMin = 0
            xMin = ((imageWidth - imageHeight) / 2.0) / imageWidth
        } else {
            width = 1
            height = imageWidth / imageHeight
            xMin = 0
            yMin = ((imageHeight - imageWidth) / 2.0) / imageHeight
        }
        return RectF(left: xMin, top: yMin, right: xMin + width, bottom: yMin + height)
    }

    /// Checks whether there are enough torso keypoints.
    /// This function checks whether the model is confident at predicting one of the
    /// shoulders/hips which is required to determine a good crop region.
    private func torsoVisible(_ keyPoints: [KeyPoint]) -> Bool {
        return
            (keyPoints[BodyPart.leftHip.position].score > minCropKeyPointScore
                || keyPoints[BodyPart.rightHip.position].score > minCropKeyPointScore)
            && (keyPoints[BodyPart.leftShoulder.position].score > minCropKeyPointScore
                || keyPoints[BodyPart.rightShoulder.position].score > minCropKeyPointScore)
    }

    /// Calculates the maximum distance from each keypoints to the center location.
    /// The function returns the maximum distances from the two sets of keypoints:
    /// full 17 keypoints and 4 torso keypoints. The returned information will be
    /// used to determine the crop size. See determineRectF for more detail.
    private func determineTorsoAndBodyDistances(
        keyPoints: [KeyPoint], targetKeyPoints: [KeyPoint], centerX: CGFloat, centerY: CGFloat) -> TorsoAndBodyDistance
    {
        let torsoJoints = [
            BodyPart.leftShoulder.position,
            BodyPart.rightShoulder.position,
            BodyPart.leftHip.position,
            BodyPart.rightHip.position,
        ]

        let maxTorsoYRange = torsoJoints.lazy.map { abs(centerY - targetKeyPoints[$0].coordinate.y) }
            .max() ?? 0.0
        let maxTorsoXRange = torsoJoints.lazy.map { abs(centerX - targetKeyPoints[$0].coordinate.x) }
            .max() ?? 0.0

        let confidentKeypoints = keyPoints.lazy.filter { $0.score < self.minCropKeyPointScore }
        let maxBodyYRange = confidentKeypoints.map { abs(centerY - $0.coordinate.y) }.max() ?? 0.0
        let maxBodyXRange = confidentKeypoints.map { abs(centerX - $0.coordinate.x) }.max() ?? 0.0

        return TorsoAndBodyDistance(
            maxTorsoYDistance: maxTorsoYRange,
            maxTorsoXDistance: maxTorsoXRange,
            maxBodyYDistance: maxBodyYRange,
            maxBodyXDistance: maxBodyXRange)
    }
}

enum MultiPoseType {
    case dynamic
    case fixed
}

enum TrackerType {
    case boundingBox
    case keypoints
    case off
}

struct InputRGBModel {
    var imageData: Data
    var shape: [Int]
}
